crates/bevy_tasks/src/task_pool.rs

```diff
@@ -0,0 +1,280 @@
+
+        let threads = (0..num_threads)
+            .map(|i| {
+                let ex = Arc::clone(&executor);
+                let flag = Arc::clone(&shutdown_flag);
+                let (p, u) = parking::pair();
+                let unparker = Arc::new(u);
+                let u = Arc::clone(&unparker);
+                // Run an executor thread.
+
+                let thread_name = if let Some(thread_name) = thread_name {
+                    format!("{} ({})", thread_name, i)
+                } else {
+                    format!("TaskPool ({})", i)
+                };
+
+                let mut thread_builder = thread::Builder::new().name(thread_name);
+
+                if let Some(stack_size) = stack_size {
+                    thread_builder = thread_builder.stack_size(stack_size);
+                }
+
+                let handle = thread_builder
+                    .spawn(move || {
+                        let ticker = ex.ticker(move || u.unpark());
+                        loop {
+                            if flag.load(Ordering::Acquire) {
+                                break;
+                            }
+
+                            if !ticker.tick() {
+                                p.park();
+                            }
+                        }
+                    })
+                    .expect("failed to spawn thread");
+
+                (handle, unparker)
+            })
+            .collect();
+
+        Self {
+            executor,
+            inner: Arc::new(TaskPoolInner {
+                threads,
+                shutdown_flag,
+            }),
+        }
+    }
+
+    /// Return the number of threads owned by the task pool
+    pub fn thread_num(&self) -> usize {
+        self.inner.threads.len()
+    }
+
+    /// Allows spawning non-`static futures on the thread pool. The function takes a callback,
+    /// passing a scope object into it. The scope object provided to the callback can be used
+    /// to spawn tasks. This function will await the completion of all tasks before returning.
+    ///
+    /// This is similar to `rayon::scope` and `crossbeam::scope`
+    pub fn scope<'scope, F, T>(&self, f: F) -> Vec<T>
+    where
+        F: FnOnce(&mut Scope<'scope, T>) + 'scope + Send,
+        T: Send + 'static,
+    {
+        let executor: &'scope multitask::Executor = unsafe { mem::transmute(&*self.executor) };
+
+        let fut = async move {
+            let mut scope = Scope {
+                executor,
+                spawned: Vec::new(),
+            };
+
+            f(&mut scope);
+
+            let mut results = Vec::with_capacity(scope.spawned.len());
+            for task in scope.spawned {
+                results.push(task.await);
+            }
+
+            results
+        };
+
+        pin_mut!(fut);
+
+        // let fut: Pin<&mut (dyn Future<Output=()> + Send)> = fut;
+        let fut: Pin<&'static mut (dyn Future<Output = Vec<T>> + Send + 'static)> =
```

Comment:
Can you add some comments here explaining what this is doing and why its safe? (including the pin_mut)

I have a general idea of how and why (lifetime hacking?), but a full description would help me (and future bevy devs).

Reply:
@lachlansneff may need to explain this but I'll give it a try.

We are creating an future of type impl Future<Output = Vec<T>> + Send + 'scope. We will need to cast away the lifetime. However, we can't use impl Trait outside of function and inherent method return types. So I'm not sure there is a way to express transmuting from that impl Trait type of 'scope lifetime to 'static lifetime. Pinning the trait makes it no longer an impl Trait type so it is possible to express the transmute in a way that ensures we are only messing with the lifetime. @lachlansneff am I way off here?

I did notice this is the only place we're using pin_mut macro so we could possibly inline that and remove the macro. I'm not sure that the #[allow(unused_mut)] is necessary here either.